plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/AtomicFUTransformer.kt

```diff
@@ -0,0 +1,595 @@
+    }
+
+    private fun transformDeclarationBody(body: IrBody?, parent: IrDeclaration, lambda: IrFunction? = null) {
+        if (body is IrBlockBody) {
+            body.statements.forEachIndexed { i, stmt ->
+                val transformedStmt = stmt.transformStatement(parent, lambda)
+                body.statements[i] = transformedStmt
+            }
+        }
+    }
+
+    private fun IrExpression.eraseAtomicConstructor(parentDeclaration: IrDeclaration, lambda: IrFunction? = null) =
+        when {
+            type.isAtomicValueType() -> getPureTypeValue().transformAtomicFunctionCall(parentDeclaration, lambda)
+            type.isAtomicArrayType() -> buildPureTypeArrayConstructor()
+            type.isReentrantLockType() -> buildConstNull()
+            else -> this
+        }
+
+    private fun IrDeclaration.transformInlineAtomicExtension(): IrDeclaration? {
+        // transform the signature of the inline Atomic* extension declaration
+        // inline fun AtomicRef<T>.foo(args) { ... } -> inline fun foo(args, getter: () -> T, setter: (T) -> Unit)
+        if (this is IrFunction &&
+            isInline &&
+            extensionReceiverParameter != null &&
+            extensionReceiverParameter!!.type.isAtomicValueType()
+        ) {
+            val newDeclaration = this.deepCopyWithSymbols(parent)
+            val oldValueParameters = valueParameters.map { it.deepCopyWithSymbols(newDeclaration) }
+            val valueParametersCount = valueParameters.size
+            val receiverValueType = extensionReceiverParameter!!.type.atomicToValueType() // T
+            val getterType = buildGetterType(receiverValueType)
+            val setterType = buildSetterType(receiverValueType)
+            newDeclaration.valueParameters = oldValueParameters + listOf(
+                buildValueParameter(newDeclaration, GETTER, valueParametersCount, getterType),
+                buildValueParameter(newDeclaration, SETTER, valueParametersCount + 1, setterType)
+            )
+            newDeclaration.extensionReceiverParameter = null
+            return newDeclaration
+        }
+        return null
+    }
+
+    private fun IrExpression.getPureTypeValue(): IrExpression {
+        require(this is IrCall && isAtomicFactory()) { "Illegal initializer for the atomic property $this" }
+        return getValueArgument(0)!!
+    }
+
+    private fun IrExpression.buildPureTypeArrayConstructor() =
+        when (this) {
+            is IrConstructorCall -> {
+                require(isAtomicArrayConstructor()) { "Unsupported atomic array constructor $this" }
+                val arrayConstructorSymbol = type.getArrayConstructorSymbol { it.owner.valueParameters.size == 1 }
+                val size = getValueArgument(0)
+                IrConstructorCallImpl(
+                    UNDEFINED_OFFSET, UNDEFINED_OFFSET,
+                    irBuiltIns.unitType, arrayConstructorSymbol,
+                    0, 0, 1
```

Comment:
If I understand correct here has to be `typeArgument` at least for `kotlin.Array<T>`

Reply:
Yeah, it's a bug, I'll pass `arrayConstructorSymbol.owner.typeParameters.size` for `typeArgumentsCount`

And what's the difference between `typeArgumentsCount` and `constructorTypeArgumentsCount` as parameters of `IrConstructorCallImpl `?

Reply:
in Java possible the following code
```java
class C<T1, T2> {
  public <CT1, CT2> C(ct: CT) { ... }
}
```
So here `typeArguments` are `<T1, T2>` and `constructorTypeArguments` are `<CT1, CT2>`